libraries/Microsoft.Bot.Connector/Authentication/ICredentialProvider2.cs

```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Threading.Tasks;
+using Microsoft.Rest;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    /// <summary>
+    /// CredentialProvider interface. This interface allows Bots to provide their own
+    /// implementation of what is, and what is not, a valid appId and password. This is 
+    /// useful in the case of multi-tenant bots, where the bot may need to call
+    /// out to a service to determine if a particular appid/password pair
+    /// is valid. 
+    /// 
+    /// For Single Tenant bots (the vast majority) the simple static providers 
+    /// are sufficient. 
+    /// </summary>
+    public interface ICredentialProvider2 : ICredentialProvider
```

Comment:
This can't get through as-is. I assume you need a different interface to maintain backward compatibility? If we need to add one more thing in the future, we'll need another interface.

Unfortunately, this is a big inconvenient, since interfaces, once published, cannot be changed for compatibility. With that in mind, could you consider creating a base class rather than an interface? That way we can add more members in the future. With this, we are digging ourselves another grave.

Alternatively, if you had verycompelling reasons for the interface, please don't name it wiht numbers.

Reply:
renamed to IServiceCredentialProvider